sync-core/src/main/java/com/cloudant/sync/query/QueryExecutor.java

```diff
@@ -0,0 +1,282 @@
+            msg += "filtering documents themselves. This will be VERY SLOW as each candidate ";
+            msg += "document is loaded from the datastore and matched against the query selector.";
+            logger.log(Level.WARNING, msg);
+        }
+
+        return new QueryResult(docIds, datastore, fields, skip, limit, matcher);
+    }
+
+    protected ChildrenQueryNode translateQuery(Map<String, Object> query,
+                                               Map<String, Object> indexes,
+                                               Boolean[] indexesCoverQuery) {
+        return (ChildrenQueryNode) QuerySqlTranslator.translateQuery(query,
+                                                                     indexes,
+                                                                     indexesCoverQuery);
+    }
+
+    protected UnindexedMatcher matcherForIndexCoverage(Boolean[] indexesCoverQuery,
+                                                       Map<String, Object> selector) {
+        return indexesCoverQuery[0] ? null : UnindexedMatcher.matcherWithSelector(selector);
+    }
+
+    private boolean validateSortDocument(List<Map<String, String>> sortDocument) {
+        if (sortDocument == null || sortDocument.isEmpty()) {
+            return true; // empty or null sort docs just mean "don't sort", so are valid
+        }
+
+        for (Map<String, String> clause: sortDocument) {
+            if (clause.size() > 1) {
+                logger.log(Level.SEVERE, "Each order clause can only be a single field");
+                return false;
+            }
+            String fieldName = (String) clause.keySet().toArray()[0];
+            String direction = clause.get(fieldName);
+            if (!direction.equalsIgnoreCase("ASC") && !direction.equalsIgnoreCase("DESC")) {
+                String msg = String.format("Order direction %s not valid, use `asc` or `desc`",
+                                           direction);
+                logger.log(Level.SEVERE, msg);
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private List<String> normaliseFields(List<String> fields) {
+        if (fields == null || fields.isEmpty()) {
+            String msg = "Projection fields array is empty, disabling project for this query";
+            logger.log(Level.WARNING, msg);
+            return null;
+        }
+
+        return fields;
+    }
+
+    private Set<String> executeQueryTree(QueryNode node, SQLDatabase db) {
+        if (node instanceof AndQueryNode) {
+            Set<String> accumulator = null;
+
+            AndQueryNode andNode = (AndQueryNode) node;
+            for (QueryNode qNode: andNode.children) {
+                Set<String> childIds = executeQueryTree(qNode, db);
+                if (accumulator == null) {
+                    accumulator = new HashSet<String>(childIds);
+                } else {
+                    accumulator.addAll(childIds);
```

Comment:
I had to make a small fix since in the previous commit version the accumulator in the else block was not actually being set to the intersection of itself and the childIds.  https://github.com/cloudant/sync-android/commit/f0ea756ce7ce7131381fc8d84c8b0f6ac16ff57c#diff-80ec4d1d12847da53428d6c14a88dbb7L208
Like mentioned before I should be able to exercise this code once the SQL translator class is capable of handling more complex queries.  Fixed in f0ea756.